Avoid parsing erroneous patterns

dolio opened the pull request:
Fixes #1731 

This also includes a new type error, in case we still parse a data pattern with ability references in it, or vice versa. It also contains actual printing for compiler bugs, which was the reason for the completely silent behavior before.

The approach I have so far actually rules out all parsing of erroneous patterns. For instance, if the example in 1731 were changed to explicitly say `CLI.input`, unison would just complain it doesn't know about a data constructor with that name (which is accurate).

This pull request is not ready quite yet, but I opened it to discuss something. There is a test case that fails currently, in `Unison.Test.TermParser` we have:

```
  match x with
    {Tuple.Cons x y} -> 1
    {Optional.Some 42 -> k} -> k 42
```

I'm not sure what this is supposed to mean, since there is no effect being matched on. Was this just a work-around to test the parser without having any ability declared?

pchiusano committed:
Add failing transcript

pchiusano committed:
Adding some comments to checkPattern

pchiusano committed:
added note on parser fix

dolio committed:
Improve some error reporting

- Add a new type error for ability/data constructors used in the
  opposite context, e.g. as erroneous patterns
- Add improved rendering of compiler bugs, and actually display
  error messages for compiler bugs.

dolio committed:
Avoid parsing data/effect patterns that don't make sense

- The way of resolving constructors in patterns wasn't taking
  into account whether a data or ability constructor was expected,
  and was therefore sometimes parsing things like variables as
  data constructors with references to ability effects.

pchiusano reviewed (approved):
LGTM.

pchiusano commented:
> The approach I have so far actually rules out all parsing of erroneous patterns.

I think it would be nice if this could be a type error rather than a syntax error, but I think you are constrained by the current AST. So your approach seems fine - I'd just delete or alter that failing test, which is just nonsense anyway.

dolio committed:
Remove test case that relies on bad parse

- This test cases expected `{ Some x -> k }` to parse, but since
  data/effect references are resolved at parse time, this isn't
  possible. Parsing effect matches is likely covered by other tests
  with legitimate abilities anyway.

dolio committed:
Only respond with type errors/bugs if the list is non-empty

dolio committed:
Transcript modifications

- I attempted to determine where the extra newlines came from,
  but I failed.

dolio committed:
Fix erroneous transcript updates.
